src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -817,6 +877,492 @@ NodeListInsertCommand(List *workerNodeList)
+	getObjectIdentityParts(address, &names, &args);
+
+	appendStringInfo(insertDistributedObjectCommand,
+					 "SELECT citus_internal_add_object_metadata(");
+
+	appendStringInfo(insertDistributedObjectCommand,
+					 "%s, ARRAY[",
+					 quote_literal_cstr(objectType));
+
+	char *name;
+	bool firstLoop = true;
+	foreach_ptr(name, names)
+	{
+		if (!firstLoop)
+		{
+			appendStringInfo(insertDistributedObjectCommand, ", ");
+		}
+		firstLoop = false;
+		appendStringInfoString(insertDistributedObjectCommand, quote_literal_cstr(name));
+	}
+
+	appendStringInfo(insertDistributedObjectCommand, "]::text[], ARRAY[");
+
+	char *arg;
+	firstLoop = true;
+	foreach_ptr(arg, args)
+	{
+		if (!firstLoop)
+		{
+			appendStringInfo(insertDistributedObjectCommand, ", ");
+		}
+		firstLoop = false;
+		appendStringInfoString(insertDistributedObjectCommand, quote_literal_cstr(arg));
+	}
+
+	appendStringInfo(insertDistributedObjectCommand, "]::text[],");
+
+	if (distributionArgumentIndex == NULL)
+	{
+		appendStringInfo(insertDistributedObjectCommand, "NULL, ");
+	}
+	else
+	{
+		appendStringInfo(insertDistributedObjectCommand, "%d, ",
+						 *distributionArgumentIndex);
+	}
+	if (colocationId == NULL)
+	{
+		appendStringInfo(insertDistributedObjectCommand, "NULL)");
+	}
+	else
+	{
+		appendStringInfo(insertDistributedObjectCommand, "%d)",
+						 *colocationId);
+	}
+
+	return insertDistributedObjectCommand->data;
+}
+
+
+/*
+ * citus_internal_add_object_metadata is an internal UDF to
+ * add a row to pg_dist_object.
+ */
+Datum
+citus_internal_add_object_metadata(PG_FUNCTION_ARGS)
+{
+	char *textType = TextDatumGetCString(PG_GETARG_DATUM(0));
+	ArrayType *nameArray = PG_GETARG_ARRAYTYPE_P(1);
+	ArrayType *argsArray = PG_GETARG_ARRAYTYPE_P(2);
+	int distributionArgumentIndexValue;
+	int colocationIdValue;
+	int *colocationId = NULL;
+	int *distributionArgumentIndex = NULL;
+
+	if (!PG_ARGISNULL(3))
+	{
+		distributionArgumentIndexValue = PG_GETARG_INT32(3);
+		distributionArgumentIndex = &distributionArgumentIndexValue;
+	}
+	if (!PG_ARGISNULL(4))
+	{
+		colocationIdValue = PG_GETARG_INT32(4);
+		colocationId = &colocationIdValue;
+	}
+
+	if (!ShouldSkipMetadataChecks())
+	{
+		/* this UDF is not allowed for executing as a separate command */
+		EnsureCoordinatorInitiatedOperation();
+	}
+
+	ObjectAddress objectAddress = GetObjectAddrIfUserOwns(textType, nameArray, argsArray);
+
+	/* Update the metadata for the given object on the node */
+	MarkObjectDistributed(&objectAddress, true);
+	UpdateFunctionDistributionInfo(&objectAddress, distributionArgumentIndex,
+								   colocationId, true);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * Get the object address if user owns it.
+ *
+ * This function is mostly copied from pg_get_object_address of the PG code. We need
+ * to copy that function to get both object type and node.
+ */
+static ObjectAddress
+GetObjectAddrIfUserOwns(char *ttype, ArrayType *namearr, ArrayType *argsarr)
```

Comment:
I'd rather call `pg_get_object_address` via `DirectFunctionCall` to copy&paste such a complex code.

Because we might need to check this code on each PG release as this might change?

Reply:
Just for reference: We prefer not use `pg_get_object_address ` because we also want to check object ownership. And, object ownership checks requires almost the same code with `pg_get_object_address`. So, we copy & paste `pg_get_object_address` and do the ownership checks inside there.